File: product-service/src/main/java/com/intexsoft/webshop/productservice/controller/SubcategoryController.java
package com.intexsoft.webshop.productservice.controller;

import com.intexsoft.webshop.productservice.dto.subcategory.SubcategoryCreateDto;
import com.intexsoft.webshop.productservice.dto.subcategory.SubcategoryDto;
import com.intexsoft.webshop.productservice.dto.subcategory.SubcategoryUpdateDto;
import com.intexsoft.webshop.productservice.service.SubcategoryService;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Positive;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

import static com.intexsoft.webshop.productservice.util.JsonUtils.getAsString;

@RestController
@RequestMapping("/subcategory")
@RequiredArgsConstructor
@Slf4j
@Validated
public class SubcategoryController {
    private final SubcategoryService subcategoryService;

    @PostMapping
    public ResponseEntity<SubcategoryDto> createSubcategory(@RequestBody @Valid
                                                            SubcategoryCreateDto subcategoryCreateDto) {
        log.info("IN: request to create a new subcategory received. Request body = {}",
                getAsString(subcategoryCreateDto));
        SubcategoryDto createdSubCategoryDto = subcategoryService.createSubcategory(subcategoryCreateDto);
        log.info("OUT: new subcategory created successfully. Response body = {}",
                getAsString(createdSubCategoryDto));
        return ResponseEntity.ok(createdSubCategoryDto);
    }

    @GetMapping
    public ResponseEntity<List<SubcategoryDto>> findAllSubcategories(@PageableDefault(size = Integer.MAX_VALUE)
                                                                     Pageable pageable) {
        log.info("IN: request to find subcategories received. Page size = {}, page number = {}",
                pageable.getPageSize(), pageable.getPageNumber());
        List<SubcategoryDto> subcategoryDtos = subcategoryService.findSubcategories(pageable);
        log.info("OUT: {} subcategories found", subcategoryDtos.size());
        return ResponseEntity.ok(subcategoryDtos);
    }

    @GetMapping("/{subcategoryId}")
    public ResponseEntity<SubcategoryDto> findSubcategoryById(@PathVariable @Positive Long subcategoryId) {
        log.info("IN: request to find a subcategory by id = {} received", subcategoryId);
        SubcategoryDto subcategoryDto = subcategoryService.findSubcategoryById(subcategoryId);
        log.info("OUT: the subcategory with id = {} found successfully. Response body = {}",
                subcategoryId, getAsString(subcategoryDto));
        return ResponseEntity.ok(subcategoryDto);
    }

    @PutMapping("/{subcategoryId}")
    public ResponseEntity<SubcategoryDto> updateSubcategory(@PathVariable @Positive Long subcategoryId,
                                                            @RequestBody @Valid SubcategoryUpdateDto subcategoryUpdateDto) {
        log.info("IN: request to update a subcategory with id = {} received. Request body = {}",
                subcategoryId, getAsString(subcategoryUpdateDto));
        SubcategoryDto updatedSubcategoryDto = subcategoryService.updateSubcategory(subcategoryId, subcategoryUpdateDto);
        log.info("OUT: the subcategory with id = {} updated successfully. Response body = {}",
                subcategoryId, getAsString(updatedSubcategoryDto));
        return ResponseEntity.ok(updatedSubcategoryDto);
    }

    @DeleteMapping("/{subcategoryId}")
    public ResponseEntity<Void> deleteSubcategoryById(@PathVariable @Positive Long subcategoryId) {
        log.info("IN: request to delete a subcategory by id = {} received", subcategoryId);
        subcategoryService.deleteSubcategoryById(subcategoryId);
        log.info("OUT: the subcategory with id = {} deleted successfully", subcategoryId);
        return ResponseEntity.noContent().build();
    }
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/controller/SubcategoryController.kt
package com.intexsoft.webshop.productservicekt.controller

import com.intexsoft.webshop.productservicekt.dto.subcategory.SubcategoryCreateDto
import com.intexsoft.webshop.productservicekt.dto.subcategory.SubcategoryDto
import com.intexsoft.webshop.productservicekt.dto.subcategory.SubcategoryUpdateDto
import com.intexsoft.webshop.productservicekt.log
import com.intexsoft.webshop.productservicekt.service.SubcategoryService
import jakarta.validation.Valid
import jakarta.validation.constraints.Positive
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.ResponseEntity
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/subcategory")
@Validated
class SubcategoryController(
    private val subcategoryService: SubcategoryService
) {
    @PostMapping
    fun createSubcategory(@RequestBody @Valid subcategoryCreateDto: SubcategoryCreateDto): ResponseEntity<SubcategoryDto> {
        log.info("IN: request to create a new subcategory received. Request body = $subcategoryCreateDto")
        val createdSubCategoryDto: SubcategoryDto = subcategoryService.createSubcategory(subcategoryCreateDto)
        log.info("OUT: new subcategory created successfully. Response body = $createdSubCategoryDto")
        return ResponseEntity.ok(createdSubCategoryDto)
    }

    @GetMapping
    fun findAllSubcategories(@PageableDefault(size = Int.MAX_VALUE) pageable: Pageable): ResponseEntity<List<SubcategoryDto>> {
        log.info(
            "IN: request to find subcategories received." +
                    " Page size = ${pageable.pageSize}, page number = ${pageable.pageNumber}"
        )
        val subcategoryDtos: List<SubcategoryDto> = subcategoryService.findSubcategories(pageable)
        log.info("OUT: ${subcategoryDtos.size} subcategories found")
        return ResponseEntity.ok(subcategoryDtos)
    }

    @GetMapping("/{subcategoryId}")
    fun findSubcategoryById(@PathVariable @Positive subcategoryId: Long): ResponseEntity<SubcategoryDto> {
        log.info("IN: request to find a subcategory by id = $subcategoryId received")
        val subcategoryDto: SubcategoryDto = subcategoryService.findSubcategoryById(subcategoryId)
        log.info("OUT: the subcategory with id = $subcategoryId found successfully. Response body = $subcategoryDto")
        return ResponseEntity.ok(subcategoryDto)
    }

    @PutMapping("/{subcategoryId}")
    fun updateSubcategory(
        @PathVariable @Positive subcategoryId: Long,
        @RequestBody @Valid subcategoryUpdateDto: SubcategoryUpdateDto
    ): ResponseEntity<SubcategoryDto> {
        log.info(
            "IN: request to update a subcategory with id = $subcategoryId received." +
                    " Request body = $subcategoryUpdateDto"
        )
        val updatedSubcategoryDto: SubcategoryDto =
            subcategoryService.updateSubcategory(subcategoryId, subcategoryUpdateDto)
        log.info(
            "OUT: the subcategory with id = $subcategoryId updated successfully." +
                    " Response body = $updatedSubcategoryDto"
        )
        return ResponseEntity.ok(updatedSubcategoryDto)
    }

    @DeleteMapping("/{subcategoryId}")
    fun deleteSubcategoryById(@PathVariable @Positive subcategoryId: Long): ResponseEntity<Unit> {
        log.info("IN: request to delete a subcategory by id = $subcategoryId received")
        subcategoryService.deleteSubcategoryById(subcategoryId)
        log.info("OUT: the subcategory with id = $subcategoryId deleted successfully")
        return ResponseEntity.noContent().build()
    }
}